lib/compiler/passes/inference-match-result.js

```diff
@@ -0,0 +1,110 @@
+"use strict";
+
+const visitor      = require("../visitor");
+const asts         = require("../asts");
+const GrammarError = require("../../grammar-error");
+
+// Inference match result of the each node. Can be:
+// -1: negative result, matching of that node always fails
+//  0: neutral result, may be fail, may be match
+//  1: positive result, always match
+function inferenceMatchResult(ast) {
+  function sometimesMatch(node) { return (node.match = 0); }
+  function alwaysMatch(node) {
+    inference(node.expression);
+
+    return (node.match = 1);
+  }
+
+  function inferenceExpression(node) {
+    return (node.match = inference(node.expression));
+  }
+  function inferenceElements(elements, forChoice) {
+    const length = elements.length;
+    let always = 0;
+    let never = 0;
+
+    for (let i = 0; i < length; ++i) {
+      const result = inference(elements[i]);
+
+      if (result > 0) { ++always; }
+      if (result < 0) { ++never;  }
+    }
+
+    if (always === length) {
+      return 1;
+    }
+    if (forChoice) {
+      return never === length ? -1 : 0;
+    }
+
+    return never > 0 ? -1 : 0;
+  }
+
+  const inference = visitor.build({
+    rule(node) {
+      let oldResult;
+      let count = 0;
+
+      // If property not yet calculated, do that
+      if (typeof node.match === "undefined") {
+        node.match = 0;
+        do {
+          oldResult = node.match;
+          node.match = inference(node.expression);
+          // 6 == 3! -- permutations count for all transitions from one match
+          // state to another.
+          // After 6 iterations the cycle with guarantee begins
+          // istanbul ignore next  This is canary test, shouldn't trigger in real life
+          if (++count > 6) {
+            throw new GrammarError(
+              "Infinity cycle detected when trying to evaluate node match result",
+              node.location
+            );
+          }
+        } while (oldResult !== node.match);
```

Comment:
When you calculate the result for a recursive rule, for example;
```peggy
start = [] start
```
On the first enter `node[name="start"].match` is `undefined`, so we set it to `0` and run the detection:
https://github.com/peggyjs/peggy/blob/36e9b07a3eab37f3342efa48f2b2a9a1dc97c7cb/lib/compiler/passes/inference-match-result.js#L49-L54

Because the rule is recursive it uses it own `match` property that we speculatively set to `0`, which may not be true (as in the example where the rule contains never matched expression so should have `match=-1`). So, when we calculate our new result we start recalculation, what will use our new result, which could lead to that we calculate the third result and so on. So, for example, our assessment about `node.match` could, for example, changing in this manner:
0 -> 1 -> 0 -> -1 -> 0 -> 1 (cycle)

I this simple example we only changed from `0` to `-1` and that result never changed again. But in more complex cases that includes several recursive rules this may not be the case.

Reply:
Please include a comment like:

```js
// an input of `start = [] start` will generate the sequence: 0 -> 1 -> 0 -> -1 -> 0 -> 1 (then cycle)
```

and I think we'll be able to remember what's going on when we see this again in a few years. :)

Reply:
That is only an example, `start = [] start` will generate sequence `0 -> -1` :smile: But I understand you intentions. You can add this yourself and merge if this is all that needed. I can do that only tomorrow